### PR TITLE
Feature/camera localization

### PR DESCRIPTION
**Pull request by Lukasz1928**

### CODE REVIEW

File: mrc/localization/color/utils/circle_points.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    def __next__(self):
+        self.current_location_vector[0] += 1
+        if self.current_location_vector[0] > self.x_range:
+            self.current_location_vector[1] -= 1
+            try:
+                self._update_x_range()
+            except ValueError:
+                raise StopIteration
+            self.current_location_vector[0] = -1 * self.x_range - 1
+        return self.__current_location()
```
Comment (zdzarsky) [inline mrc/localization/color/utils/circle_points.py]: No newline 😢

File: mrc/localization/color/utils/color_converter.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import cv2
+
+from mrc.shared.exceptions.exceptions import ColorEncodingNotSupportedException
+
+
+def hsv2grayscale(image):
```
Comment (zdzarsky) [inline mrc/localization/color/utils/color_converter.py]: It seems we have a beautiful place for such function at mrc/color/utils.py

File: requirements.txt [hunk 3]
```diff
@@ -1,4 +1,3 @@
 opencv-python
 numpy
 scikit-learn
-pytest
```
Comment (zdzarsky) [inline requirements.txt]: image

File: mrc/localization/color/utils/circle_points.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, centre, radius):
+        self.centre = centre
+        self.radius = radius
+        self.current_location_vector = [0, radius + 1]
+        self.x_range = 0
+
+    def _update_x_range(self):
+        if self.current_location_vector[1] in [self.radius, -self.radius]:
+            self.x_range = 0
+        elif self.current_location_vector[1] >= 0:
```
[the remark this answers is not in the document]
Comment (zdzarsky) [reply mrc/localization/color/utils/circle_points.py]: Roses are red.
Violets are blue.
Omae wa mo
Shindeiru.
🤷‍♂️

File: tests/test_utils/read_image.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import cv2
+
+
+def read_image(name):
+    image = cv2.imread('tests/resources/{}'.format(name))
```
Comment (zdzarsky) [inline tests/test_utils/read_image.py]: You said something about pngs here.

File: mrc/shared/exceptions/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from mrc.shared.exceptions.mrc_exception import MRCException
+
+
+class IncorrectShapeException(MRCException):
+    pass
```
Comment (zdzarsky) [inline mrc/shared/exceptions/exceptions.py]: Inits with msg would be awesome.

Comment (Lukasz1928) [reply mrc/shared/exceptions/exceptions.py]: msg is implemented by default ;)

File: tests/test_utils/read_image.py (hunk 5 above)
Comment (Lukasz1928) [reply tests/test_utils/read_image.py]: done

File: mrc/localization/camera/utils/connected_components_detector.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+import math
+import cv2
+from mrc.utils.maths import points_2d_sqr_distance, point_in_rectangle, rescale_rectangle
+
+
+class ConnectedComponentsDetector:
+    def __init__(self):
```
Comment (zdzarsky) [inline mrc/localization/camera/utils/connected_components_detector.py]: - [ ] ?

File: mrc/localization/camera/utils/connected_components_detector.py [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            centroids_list.pop(background_component_index)
+        if remove_irrelevant and remove_background:
+            irrelevant_indexes = self._get_irrelevant_components_indexes(stats_list, centroids_list, 500, 25)
+            stats_list = [stats_list[i] for i in range(len(stats_list)) if i not in irrelevant_indexes]
+            centroids_list = [centroids_list[i] for i in range(len(centroids_list)) if i not in irrelevant_indexes]
+        return stats_list, centroids_list
+
+    def _get_background_component_index(self, image, stats, centroids):
+        image_centre = (image.shape[0] / 2, image.shape[1] / 2)
+        areas = [s[cv2.CC_STAT_HEIGHT] * s[cv2.CC_STAT_WIDTH] for s in stats]
```
Comment (zdzarsky) [inline mrc/localization/camera/utils/connected_components_detector.py]: - [ ] Why not`: areas = [x *y for x,y,_,_,_ in stats]`

Comment (Lukasz1928) [reply mrc/localization/camera/utils/connected_components_detector.py]: What if it changes with OpenCV version?

**Pull request merged**